### ISSUE TITLE
Proposal: Allow internal interface members

### ISSUE THREAD

**Issue by JeffCyr** [dotnet/roslyn#30452]
# Context
When using interfaces, you may want to allow the implementation of an interface publicly but restrict the usage of some members to the assembly that defined the interface.

See the discussion in this issue about IThreadPoolWorkItem:
https://github.com/dotnet/corefx/issues/32485

```c#
public interface IThreadPoolWorkItem
{
    void Execute();
}
```
The problem is that we want a type to implement `IThreadPoolWorkItem.Execute` (e.g. `System.Threading.Tasks.Task`), but prevent code outside corefx from calling directly the `Execute` method because it would yield undefined behavior.

# Proposal

Allow the `internal` keyword on interface members. Internal members can be implemented but can't be used outside of the assembly where the interface is defined.

```c#
public interface IFoo
{
    internal void Bar();
}

public class Foo : IFoo
{
    // An internal interface member can only be explicitly implemented
    void IFoo.Bar()
    {
    }
}

Foo o = new Foo();

// Only allowed if IFoo is declared in the current assembly
o.Bar();
```

cc @stephentoub

**Comment by svick** [comment]
You seem to be suggesting a language change. Per [the Readme](https://github.com/dotnet/roslyn#language-design-discussion), those belong in the [csharplang repo](https://github.com/dotnet/csharplang/).

**Comment by jaredpar** [comment]
Thanks for the suggestion. As @svick noted though, this is a language suggestion and hence more suited for the csharplang repository.

**Comment by stephentoub** [comment]
Is this just a language thing?  i.e. does the CLR actually support that at the IL level?

**Comment by jaredpar** [comment]
Yes it is supported at the IL level.

**Comment by gafter** [comment]
We are already doing this as part of the *default interface implementations* feature.  See https://github.com/dotnet/csharplang/issues/52

**Comment by sonnemaf** [comment]
The internal interface members don't work as expected (for me).

Take a look at this ClassLibrary. It contains the IDisplayable interface with 2 internal members. The Display(string, int) with some inplementation and a Foo() method with no implementation.

```cs
using System;

namespace ClassLibrary1 {

    public interface IDisplayable {

        void Display(string text);

        internal void Display(string text, int times) {
            for (int i = 0; i < times; i++) {
                this.Display(text);
            }
        }

        internal int Foo();
    }

    public class Report : IDisplayable {

        public void Display(string text) => Console.WriteLine("Report " + text);

        int IDisplayable.Foo() => 42;
    }
}
```

In the class Report I had to implement the Foo() method explicitly. This is ok.

If I create a second project/assembly which references this ClassLibrary the problem arises.

```cs
using System;
using ClassLibrary1;

namespace ConsoleApp1 {

    class Program {
        static void Main(string[] args) {
            IDisplayable d = new Report();
            d.Display("abc");
            //d.Display("abc", 10); // Won't compile, this is great
        }
    }

    class Demo : IDisplayable {
        public void Display(string text) => Console.WriteLine("Demo " + text);
    }
}
```

The Display(string, int) method can't be called because it is internal. This is great!

The Demo class can't implement the IDisplayable interface though. Compiler error **CS0535 'Demo' does not implement interface member 'IDisplayable.Foo()'**

![image](https://user-images.githubusercontent.com/9335559/96102459-d5f00800-0ed6-11eb-9cbf-0cb69a15528f.png)

When I add an explicit implementation for Foo I get this error.

![image](https://user-images.githubusercontent.com/9335559/96102578-f750f400-0ed6-11eb-9e4c-4782370223f4.png)

It is of course also not fixed if I make the Foo method public. 

![image](https://user-images.githubusercontent.com/9335559/96102852-472fbb00-0ed7-11eb-8932-2f76c7c85e9c.png)

@gafter I don't think it is correctly implemented in dotnet/csharplang#52
